### PR TITLE
Spectrum code

### PR DESCRIPTION
**Pull request by bwyss**
This PR extends the HRDE application to support single and multiple (per site) hazard spectrum curves. 
The spectrum curves were not able to follow the same paradigm as the UHS or hazard curves because each spectrum curve has unique values for the x and y axis, while hazard and UHS curves share a single x-axis array. 
Because of this, and the fact that a spectrum curve should use 200 +- nodes to describe a single curve, it was decided to compute each curve in the client using 5 parameters and a vector from 0 to 5 with steps of 0.05. 

As of the time of this PR is being opened these features are 'live' in the OQP dev environment. 

1. Click load hazard data button
2. From the Hazard Data dialog, access the Europe model
3. Once modes are loaded into the menu scroll to the bottom of the dialog to "Hazard Spectrum Code"
4. Select and add the mode
5. Click a point on the map to render spectrum curve(s)
6. download CSV of curve(s)

### CODE REVIEW

**Comment by nastasi-oq** [comment]
Comments after test: 
 * Download curve not styled
 * The name of the downloaded file could be related to the name of the curve
 * The pop-up window increase it's dimension at each new click

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js [hunk 1]
```diff
@@ -0,0 +1,439 @@
...
+            .attr("r", 2.5)
+            .style("fill", color)
+            .on("mouseover", function() {
+                d3.select(this)
+                    .attr('r', 6)
+                    .text(circleX + ", " + circleY)
+                    .style("fill", "gray");
+                var circleX = d3.select(this.__data__[0]);
+                circleX = circleX.toString();
+                circleX = circleX.split(","[0]);
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js]: ```","[0] == ","```  => True
so why use it ?

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js [hunk 2]
```diff
@@ -0,0 +1,439 @@
...
+        .text("(Lon/Lat): "+lon+", "+lat);
+
+    textBottom = svg.append("text")
+        .attr("x", 0)
+        .attr("y", 315)
+        .attr("dy", ".35em")
+        .style("font-size","11px")
+        .text('');
+
+    $('#chartDialog').append('<div id="downloadCurve"><font color="blue">Download Curve</font></div>');
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js]: Please, ```font``` is not supported in html5 ... please use ```<p style="color: blue;">Download Curve</p>``` instead or equivalent.

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js [hunk 3]
```diff
@@ -0,0 +1,439 @@
...
+        var csvHeader = selectedCurves;
+        var csvData = [];
+        csvData = csvData.concat(csvHeader);
+        csvData.push("investigationTime");
+        csvData.push("poE");
+        csvData.push("lon");
+        csvData.push("lat");
+        csvData = JSON.stringify(csvData);
+        var lineBreak = "lineBreak";
+        csvData = csvData.concat(lineBreak);
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js]: ```concat``` method concatenates the instance array with arrays passed as arguments.
At this line and many others below the ```push``` method is more correct and efficient to just add a new element at the end of array (without reassign it).

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js [hunk 4]
```diff
@@ -0,0 +1,439 @@
...
+        }
+
+        csvData = csvData.concat(',');
+        csvData = csvData.concat(invest_time);
+        csvData = csvData.concat(',');
+        csvData = csvData.concat(poe);
+        csvData = csvData.concat(',');
+        csvData = csvData.concat(lon);
+        csvData = csvData.concat(',');
+        csvData = csvData.concat(lat);
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js]: all these lines could be replaced with 
```csvData.push(',');
csvData.push(invest_time);
...
```

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        csvData = csvData.concat('"');
+        csvData = csvData.concat(probArray);
+        csvData = csvData.concat('","');
+        csvData = csvData.concat(layerIml);
+        csvData = csvData.concat('",');
+        csvData = csvData.concat(AppVars.layerInvestigationTime);
+        csvData = csvData.concat(',');
+        csvData = csvData.concat(lng);
+        csvData = csvData.concat(',');
+        csvData = csvData.concat(lat);
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js]: Same comment as above.

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js (hunk 3 above)
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js]: why not use ```csvData += lineBreak``` construct ?

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js (hunk 4 above)
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_mixed.js]: In this case:
```csvData += ',' + invest_time + ',' + poe + .... ```

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js [hunk 6]
```diff
@@ -0,0 +1,200 @@
...
+    });
+
+    var h = $('#chartDialog').height();
+    h = h + 20;
+
+    $('#chartDialog').css({'height': h+'px'});
+
+    var prob, iml;
+
+    if (curveType === 'spectrum') {
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js]: ```==``` isn't enough ?

File: openquakeplatform/openquakeplatform/static/js/hrde_input_hazard_chart.js [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+            .replace(/\[/g, '')
+            .replace(/\]/g, '')
+            .replace(/","/g, ',')
+            .replace(/"/g, '');
+
+        for (var k in selectedCurves) {
+            curve_name = selectedCurves[k];
+            var curveValue = mfdsJsonObj[curve_name];
+
+            csvData += curve_name;
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_input_hazard_chart.js]: Are we sure that curve_name don't includes single or double quote that must be escaped to fit in a valid CSV file ?

File: openquakeplatform/openquakeplatform/static/js/hrde_loss_curve_chart.js [hunk 8]
```diff
@@ -0,0 +1,274 @@
...
+            .attr("r", 2.5)
+            .style("fill", color)
+            .on("mouseover", function() {
+                d3.select(this)
+                    .attr('r', 6)
+                    .text(circleX + ", " + circleY)
+                    .style("fill", "gray");
+                var circleX = d3.select(this.__data__[0]);
+                circleX = circleX.toString();
+                circleX = circleX.split(","[0]);
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_loss_curve_chart.js]: Same as above.

File: openquakeplatform/openquakeplatform/static/js/hrde_loss_curve_chart.js [hunk 9]
```diff
@@ -0,0 +1,274 @@
...
+            "orange",
+            "red",
+            "sandybrown",
+            "yellowgreen",
+            "darksalmon",
+            "lightseagreen",
+            "skyblue"
+        ];
+
+        var gray = "A0A0A0";
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_loss_curve_chart.js]: No ```#``` symbol before ```A0A0A0``` ?

File: openquakeplatform/openquakeplatform/static/js/hrde_loss_curve_chart.js [hunk 10]
```diff
@@ -0,0 +1,274 @@
...
+        .text('Loss Curve');
+
+    textTop = svg.append("text")
+        .attr("x", 0)
+        .attr("y", -15)
+        .attr("dy", ".35em")
+        .attr("font-size","11px")
+        .text('');
+
+    $('#chartDialog').append('<div id="downloadCurve"><font color="blue">Download Curve</font></div>');
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_loss_curve_chart.js]: Same as above.

File: openquakeplatform/openquakeplatform/static/js/hrde_spectrum_chart_mixed.js [hunk 11]
```diff
@@ -0,0 +1,215 @@
...
+        .attr("dy", ".35em")
+        .attr("font-size","12px")
+        .text('Point coordinates (lon, lat [in DD]): '+lng+', '+lat);
+    textBottom = svg.append('text')
+        .attr('x', 0)
+        .attr('y', 335)
+        .attr('dy', '.35em')
+        .text('');
+
+    $('#chartDialog').append('<div id="downloadCurve"><font color="blue">Download Curve</font></div>');
```
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_spectrum_chart_mixed.js]: Same as above.

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js (hunk 6 above)
Comment (bwyss) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js]: please see: http://stackoverflow.com/questions/359494/does-it-matter-which-equals-operator-vs-i-use-in-javascript-comparisons

File: openquakeplatform/openquakeplatform/static/js/hrde_input_hazard_chart.js (hunk 7 above)
Comment (bwyss) [inline openquakeplatform/openquakeplatform/static/js/hrde_input_hazard_chart.js]: we are sure

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js (hunk 6 above)
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js]: In this case why not in all the other check occurencies like:
```if (curveType == 'uhs' || curveType == 'spectrum') {```

File: openquakeplatform/openquakeplatform/static/js/hrde_input_hazard_chart.js (hunk 7 above)
Comment (nastasi-oq) [inline openquakeplatform/openquakeplatform/static/js/hrde_input_hazard_chart.js]: ok

File: openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js (hunk 6 above)
Comment (bwyss) [inline openquakeplatform/openquakeplatform/static/js/hrde_hazard_chart_single.js]: "The identity (===) operator behaves identically to the equality (==) operator except no type conversion is done"

**Comment by nastasi-oq** [comment]
LGTM

**Pull request merged**